src/services/pallets/PalletsErrorsService.ts

```diff
@@ -0,0 +1,298 @@
+			errorItem: errorItemId,
+			metadata: normalizedErrorItemMeta,
+		};
+	}
+
+	async fetchErrors(
+		historicApi: ApiDecoration<'promise'>,
+		{ hash, palletId, onlyIds }: IFetchPalletArgs & { onlyIds: boolean }
+	): Promise<IPalletErrors> {
+		const metadata = historicApi.registry.metadata;
+		const [palletMeta, palletMetaIdx] = this.findPalletMeta(
+			metadata,
+			historicApi,
+			palletId
+		);
+
+		const { number } = await this.api.rpc.chain.getHeader(hash);
+		const errors = historicApi.errors[palletMeta.name.toString().toLowerCase()];
+
+		let items: [] | ErrorMetadataLatest[] | Text[];
+		if (palletMeta.errors.isNone) {
+			items = [];
+		} else if (onlyIds) {
+			items = Object.entries(errors).map(
+				(errorItem) => errorItem[0] as unknown as Text
+			);
+		} else {
+			items = [] as ErrorMetadataLatest[];
+			for (const [, value] of Object.entries(errors)) {
+				const item = value.meta;
+				items.push(item);
+			}
+		}
+
+		return {
+			at: {
+				hash: hash,
+				height: number.unwrap().toString(10),
+			},
+			pallet: stringCamelCase(palletMeta.name),
+			palletIndex: palletMetaIdx,
+			items,
+		};
+	}
+
+	/**
+	 * Normalize error item metadata by running it through `sanitizeNumbers` and
+	 * converting the docs section from an array of strings to a single string
+	 * joined with new line characters.
+	 *
+	 * @param errorItemMeta polkadot-js ErrorEntryMetadataV12
+	 */
+	private normalizeErrorItemMeta(
+		errorItemMeta: ErrorMetadataLatest
+	): ISanitizedErrorItemMetadata {
+		const normalizedErrorItemMeta = sanitizeNumbers(
+			errorItemMeta
+		) as unknown as ISanitizedErrorItemMetadata;
+
+		normalizedErrorItemMeta.docs = this.sanitizeDocs(errorItemMeta.docs);
+
+		return normalizedErrorItemMeta;
+	}
+
+	/**
+	 * Find the error item's metadata within the pallets's metadata.
+	 *
+	 * @param palletMeta the metadata of the pallet that contains the error item
+	 * @param errorId name of the error item in camel or pascal case
+	 */
+	private findErrorItemMeta(
+		palletMeta: PalletMetadataV14,
+		errorItemId: string,
+		errors: ModuleErrors<'promise'>
+	): PalletErrorMetadataV14 {
+		if (palletMeta.errors.isNone) {
+			throw new InternalServerError(
+				`No error items found in ${palletMeta.name.toString()}'s metadata`
+			);
+		}
+
+		let errorItemMetaIdx = -1;
+		for (const [, value] of Object.entries(errors)) {
+			if (value.meta.name.toLowerCase() === errorItemId.toLowerCase()) {
+				errorItemMetaIdx = value.meta.index.toNumber();
+			}
+		}
+
+		if (errorItemMetaIdx === -1) {
+			throw new InternalServerError(
+				`Could not find error item ("${errorItemId}") in metadata. Error item names are expected to be in camel case, e.g. 'errorItemId'`
+			);
+		}
+
+		return Object.entries(errors)[
+			errorItemMetaIdx
+		] as unknown as PalletErrorMetadataV14;
+	}
+
+	/**
+	 * Find a pallet's metadata info.
+	 *
+	 * @param palletId identifier for a FRAME pallet as a pallet name or index.
+	 */
+	private findPalletMeta(
```

Comment:
How much of this is reusable, if we are going to add multiple other pallet endpoints? Could this be a part of the AbstractService? or maybe we add an AbstractPalletService?

Reply:
I think adding an abstract pallet service makes sense. Ill work on adding that to see what can be reused between these especially since we are adding more pallet endpoints